File: JAVA/OBJECT ORIENTED PROGRAMMING/APPLICATIONS PROJECTS/src/PhoneBookApp/PhoneBook.java
package PhoneBookApp;
import java.util.ArrayList;
import java.util.List;


public class PhoneBook{
    private List<Contact> contacts;

    public PhoneBook(){
        this.contacts = new ArrayList<Contact>();
    }
    public void addContact(String firstName, String lastName, String phoneNumber){
        if (isValidPhoneNumber(phoneNumber)) {
            this.contacts.add(new Contact(firstName, lastName, phoneNumber));
        } else {
            System.out.println("Invalid phone number");
        }
    }
    public void removeContact(String firstName, String lastName, String phoneNumber){
        for(Contact contact : this.contacts){
            if(contact.getFirstName().equals(firstName) && contact.getLastName().equals(lastName)
                    && (contact.getPhoneNumber().equals(phoneNumber))){
                this.contacts.remove(contact);
            }
        }
    }
    public List<Contact> getContacts(){
        return this.contacts;
    }
    public Contact findContactByFirstName(String firstName){
        for (Contact contact : this.contacts){
            if (contact.getFirstName().equals(firstName)){
                return contact;
            }
        }
        return null;
    }
    public Contact findContactByLastName(String lastName){
        for (Contact contact : this.contacts){
            if (contact.getLastName().equals(lastName)){
                return contact;
            }
        }
        return null;
    }
    public Contact findContactByNumber(String number){
        for (Contact contact : this.contacts){
            if (contact.getPhoneNumber().equals(number)){
                return contact;
            }
        }
        return null;
    }
    public String editContact(String newFirstName, String newLastName, String number){
        for (Contact contacts : this.contacts){
            if (contacts.getPhoneNumber().equals(number)){
                    contacts.setFirstName(newFirstName);
                    contacts.setLastName(newLastName);
            }
        }
        return null;
    }
    public boolean isValidPhoneNumber(String phoneNumber){
        return (phoneNumber.startsWith("080") ||
                phoneNumber.startsWith("081") ||
                phoneNumber.startsWith("090") ||
                phoneNumber.startsWith("091") ||
                phoneNumber.startsWith("070") ||
                phoneNumber.startsWith("+234") && phoneNumber.length() == 13) &&
                phoneNumber.length() == 11;
    }
}

File: PYTHON/PHONEBOOK/src/phone_book.py
# import List

from PHONEBOOK.src.contact import Contact


class PhoneBook:
    def __init__(self):
        self.contacts = []

    def add_contact(self, contact):
        if not self.is_phone_number_valid(contact.phone_number):
            raise ValueError("Invalid phone number")

        self.contacts.append(contact)
        return self.contacts


    def remove_contact(self, contact: Contact):
        for contact in self.contacts:
            self.contacts.remove(contact)

    def get_contacts(self):
        return self.contacts

    def find_contact_by_first_name(self,  first_name):
        for contact in self.contacts:
            if contact.first_name == first_name:
                return contact
        raise ValueError("Contact not found")

    def find_contact_by_last_name(self, second_name):
        for contact in self.contacts:
            if second_name == contact.last_name:
                return contact
        raise ValueError("Invalid phone number")

    def find_contact_by_phone_number(self, number):
        for contact in self.contacts:
            if number == contact.phone_number:
                return contact
        return None

    def edit_contact(self, first_name: str, new_first_name:str, last_name: str, new_last_name:str, phone_number: str):
        contact = self.find_contact_by_first_name(first_name)
        if contact:
            contact.first_name = new_first_name
            contact.last_name = new_last_name
            contact.phone_number = phone_number
            return contact
        return None

    @staticmethod
    def is_phone_number_valid(phone_number:str):
        global len
        if phone_number.startswith(("080", "081", "090", "091", "070")) and len(phone_number) == 11:
            return True
        elif phone_number.startswith("+243") and len(phone_number) == 13:
            return True
        else:
            return False

phone_book = PhoneBook()
phone_book.add_contact(Contact("Oba", "Joy", "08011111111"))

print(phone_book.get_contacts()[0].first_name)
